### ISSUE TITLE
[FEATURE REQUEST] Visualization of same notes at same time without overlap

### ISSUE THREAD

**Issue by TS53** [spessasus/SpessaSynth#89]
When two or more voices play the same note at same time, only one of the voices is shown and the voice(s) below are hidden.
The following is an example:
[ThreeVoices.zip](https://github.com/user-attachments/files/18238876/ThreeVoices.zip)

If it's not too much work to implement this, it would be nice to show all voices instead of only one.
I can imagine the following visualizations:
- create the note with the same width like always but built from all colors of the voices that appear at this time at this note.
For example, when two voices have this note at this time, make it 50 % color1 and 50 % color 2. With three notes 33 % of each color.
- stripe the note like above but [diagonal](https://css-tricks.com/diagonal-stripes-wipe-animation/)

**Comment by spessasus** [comment]
I am against your suggestion. The voices in your example appear to have the same length, so the color which is show is not determined. Spessasynth renders the shortest notes at the front and longest at the back, like every other visualization I am aware of that handles overlapping notes (Synthesisia or Midiano for example) 

For example if blue note lasts for 5 seconds, and then a red note lasts for a second somewhere in the middle, the end result will be a blue note with the red note drawn over it in the place where it gets pressed. Also stripes would be tricky to implement as it's not CSS, but canvas. Not to mention that is quite rare that two exactly the same notes play at the exact same time.

Here's how Midiano handles overlapping notes:
![image](https://github.com/user-attachments/assets/0a7ace35-ee45-4bb0-9c0e-8293f49fb5cf)

And here's Spessasynth:
![image](https://github.com/user-attachments/assets/d38b1e60-eb4c-4436-b800-6f5f1193bfea)
As you can see, they are the same and quite clean IMO.

**Comment by TS53** [comment]
> The voices in your example appear to have the same length

Exactly! I made them look like they have the same length while in fact they do not! To show how difficult it is to see when and how long the underlying note is actually played.

![same length](https://github.com/user-attachments/assets/60dbe383-f85c-4f5f-8375-bd65269c78d2)


> like every other visualization I am aware

Yes, I also have not seen a single one that can do this and wondered why.

> Not to mention that is quite rare that two exactly the same notes play at the exact same time.

Unfortunately, it's quite often, ca. 50 % of my MIDIs have this problems and they are from different composers.

> Also stripes would be tricky to implement as it's not CSS, but canvas.

Ok, that's unfortunate. Maybe a different way of visualization appears that makes it more obvious. Also when the shorter note is on top, which is good, it is not clear, if at the same time the note below is played.

**Comment by TS53** [comment]
Maybe just put in a small  circle of each color that is played simultanesly in the shortest note. This would be simple to render I think and would make it clear. But I'm not 100 % sure how good this will look in praxis.

**Comment by spessasus** [comment]
Can you maybe take a screenshot and draw a visual representation of your ideas?

**Comment by TS53** [comment]
The first note (orange) looks like this currently:

![first note](https://github.com/user-attachments/assets/3f72d1f4-f699-47bd-b295-524ea54902d0)

It hides the red and yellow notes. One of them has the same length. Nonetheless, the color orange is shown here as the only one.
I think, in the orange note it should already made clear, that there are more notes behind that are currently played and are hidden. Like this one (maybe not the best way, but a first idea):

![first note with hidden notes](https://github.com/user-attachments/assets/c991c94d-8e9f-49cf-a233-b94209382e70)

When a note that is displayed on top hides an other note which finishes earlier, the small rectangle of this shorter note should only get displayed at the position of the note.

In the example above two colors are chosen for the small rectangles because two notes are hidden. If it is only one note then only one color is drawn. If it is three notes, then three colors in an alternating pattern.

This is my first idea of solving this.

**Comment by TS53** [comment]
Another idea would be to this note by a sequences of smaller rectangles with rounded corners in the alternating pattern but not on top of the orange note. Instead, there is no main note (here orange was chosen as main note). The alternating blocks of rounded rectangles form the note and all blocks get the shadow.

**Comment by spessasus** [comment]
> You should also implement note-level oscilloscope views, like in _this_ https://youtu.be/VXFNAWD-194
> 
> I've also done videos that combine this trick with per-instrument oscilloscope views like this one https://youtu.be/8f_QI-Iz7dE
> 
> And I've even combined it with vocals https://youtu.be/_giDYtGG27A

First of all, this is a completely different feature request, not related to this one. Please open another issue.

Secondly, I am against this for a few reasons:
1. Performance
    - SpessaSynth already has oscilloscopes, 16 of them. And on my, modern and rather fast phone, it already significantly decreases the performance. 
    - Having more oscilloscopes will absolutely kill it the performance.
    - Not to mention, that the program you've probably used (it was called corrscope, I think) renders this offline.
    -  I'm sure that this render took longer than realtime, as using the real-time preview already lags my computer on high framerates with just 16 waveforms, which proves that it's not suitable for real-time visualizations.
    -  And also, the algorhitm corrscope uses for triggering is very computationally expensive, way more than the simple zero-crossing approach SpessaSynth currently uses

2. Visibility
   - For every voice? With LiveHQ soundfont and complex MIDIs, the voice count easily hits 300 and more. Do you expect  *600* oscilloscopes (2 for each voice like in your video)?
   - Another reminder: it's *real-time.* The synth *does not know* how many voices it is going to play. So should it always have `VOICE_CAP` amount oscilloscopes ready? And if so, the voice cap is 350 by default. So 700 oscilloscopes.

Sorry, but this simply isn't applicable for a real-time application.

**Comment by stgiga** [comment]
> > You should also implement note-level oscilloscope views, like in _this_ https://youtu.be/VXFNAWD-194
> > I've also done videos that combine this trick with per-instrument oscilloscope views like this one https://youtu.be/8f_QI-Iz7dE
> > And I've even combined it with vocals https://youtu.be/_giDYtGG27A
> 
> First of all, this is a completely different feature request, not related to this one. Please open another issue.
> 
> Secondly, I am against this for a few reasons:
> 
> 1. Performance
>    
>    * SpessaSynth already has oscilloscopes, 16 of them. And on my, modern and rather fast phone, it already significantly decreases the performance.
>    * Having more oscilloscopes will absolutely kill it the performance.
>    * Not to mention, that the program you've probably used (it was called corrscope, I think) renders this offline.
>    * I'm sure that this render took longer than realtime, as using the real-time preview already lags my computer on high framerates with just 16 waveforms, which proves that it's not suitable for real-time visualizations.
>    * And also, the algorhitm corrscope uses for triggering is very computationally expensive, way more than the simple zero-crossing approach SpessaSynth currently uses
> 2. Visibility
>    
>    * For every voice? With LiveHQ soundfont and complex MIDIs, the voice count easily hits 300 and more. Do you expect  _600_ oscilloscopes (2 for each voice like in your video)?
>    * Another reminder: it's _real-time._ The synth _does not know_ how many voices it is going to play. So should it always have `VOICE_CAP` amount oscilloscopes ready? And if so, the voice cap is 350 by default. So 700 oscilloscopes.
> 
> Sorry, but this simply isn't applicable for a real-time application.

Sorry, I thought it was relevant, and honestly the speed penalty is a concern. My apologies.

**Comment by spessasus** [comment]
I personally am against this feature request, as I've gotten used to the current look and think that it looks good, not to mention that implementing this idea would be quite complex, especially making it perform well.

However, you're welcome to create a pull request and I'll add it as a toggleable setting for the renderer.

**Comment by TS53** [comment]
Ok, I understand your points. Unfortunately, my frontend skilly are limited, so I will close this.

**Comment by TS53** [comment]
Another idea is to put the louder voice on top of the others. Using this technique, it is possible to easily control which voice should be on top using the Synthesizer Controller. When all Voices have the same loudness, the same method like now can get used.
